Add bidder name key support

camrice opened the pull request:
Original issue and more info for this change: https://github.com/prebid/prebid-server/issues/1458

hhhjort commented on exchange/exchange_test.go:
```diff
@@ -1638,6 +1638,114 @@ func TestNoCategoryDedupe(t *testing.T) {
+	bidderName1 := openrtb_ext.BidderName("appnexus1")
+
+	seatBid2 := pbsOrtbSeatBid{innerBids2, "USD", nil, nil}
+	bidderName2 := openrtb_ext.BidderName("appnexus2")
```
I might suggest instead of `appnexus1` and `appnexus2` you either go with `bidder1` and `bidder2` or just choose two different existing bidders. This suggests that there might be something special about having `appnexus` in your bidder name.

hhhjort replied:
@camrice : Did you see this comment?

camrice replied:
Sorry @hhhjort ! I just saw this comment. I updated the bidder names to be `bidder1/2`.

hhhjort reviewed (approved):
LGTM

guscarreon commented on endpoints/openrtb2/video_auction.go:
```diff
@@ -624,6 +624,7 @@ func createBidExtension(videoRequest *openrtb_ext.BidRequestVideo) ([]byte, erro
 		IncludeBrandCategory: inclBrandCat,
 		DurationRangeSec:     durationRangeSec,
 		IncludeBidderKeys:    true,
+		AppendBidderNames:    videoRequest.AppendBidderNames,
```
Given that `IncludeBidderKeys` is set to `true` in line 626 we could probably end up with a targeting that appends `_bidderName` in both the key and the value of the target. Something like:
```
  `"hb_pb_cat_dur_appnxs": "15.00_394_30s_appnxs"`
```
Instead of what issue #1458 specified:
```
  `"hb_pb_cat_dur": "15.00_394_30s_appnxs"`
```
Did you intend this? If not, should `IncludeBidderKeys` and `AppendBidderNames` be mutually exclusive?

guscarreon replied:
Unless I'm wrong, target key `"hb_pb_cat_dur` could get appended `_appnxs` in the `targData.setTargeting()` function that gets called inside `HoldAuction()` in line 216 below:
```
107 func (e *exchange) HoldAuction(ctx context.Context, bidRequest *openrtb.BidRequest, usersyncs IdFetcher, labels pbsmetrics.Labels, account *config.Account, categoriesFetcher *stored_requests.CategoryFetcher, debugLog *DebugLog)
108
109 +-- 63 lines: requestExt, err := extractBidRequestExt(bidRequest)-----------------------------------------------------------------------------
172     adapterBids, adapterExtra, anyBidsReturned := e.getAllBids(auctionCtx, cleanRequests, aliases, bidAdjustmentFactors, blabels, conversions)
173
174     var auc *auction = nil
175     var bidResponseExt *openrtb_ext.ExtBidResponse = nil
176     if anyBidsReturned {
177
178 +-- 15 lines: var bidCategory map[string]string-----------------------------------------------------------------------------------------------
193
194         if targData != nil {
195             auc.setRoundedPrices(targData.priceGranularity)
196 +-- 20 lines: if requestExt.Prebid.SupportDeals {---------------------------------------------------------------------------------------------
216             targData.setTargeting(auc, bidRequest.App != nil, bidCategory)
217
218             // Ensure caching errors are added if the bid response ext has already been created
219             if bidResponseExt != nil && len(cacheErrs) > 0 {
220                 bidderCacheErrs := errsToBidderErrors(cacheErrs)
221                 bidResponseExt.Errors[openrtb_ext.PrebidExtKey] = append(bidResponseExt.Errors[openrtb_ext.PrebidExtKey], bidderCacheErrs...)
222             }
223         }
224
225     }
226 +-- 19 lines: if !anyBidsReturned {-----------------------------------------------------------------------------------------------------------
245     return e.buildBidResponse(ctx, liveAdapters, adapterBids, bidRequest, adapterExtra, auc, bidResponseExt, cacheInstructions.returnCreative, errs)
246 }
exchange/exchange.go
```
And then lines 76 and 85:
```
 37 func (targData *targetData) setTargeting(auc *auction, isApp bool, categoryMapping map[string]string) {
 38     for impId, topBidsPerImp := range auc.winningBidsByBidder {
 39         overallWinner := auc.winningBids[impId]
 40         for bidderName, topBidPerBidder := range topBidsPerImp {
 41 +-- 34 lines: isOverallWinner := overallWinner == topBidPerBidder------------------------------------------------
 75             if len(categoryMapping) > 0 {
 76                 targData.addKeys(targets, openrtb_ext.HbCategoryDurationKey, categoryMapping[topBidPerBidder.bid.ID], bidderName, isOverallWinner)
 77             }
 78
 79             topBidPerBidder.bidTargets = targets
 80         }
 81     }
 82 }
 83
 84 func (targData *targetData) addKeys(keys map[string]string, key openrtb_ext.TargetingKey, value string, bidderName openrtb_ext.BidderName, overallWinner bool) {
 85     if targData.includeBidderKeys {
 86         keys[key.BidderKey(bidderName, MaxKeyLength)] = value
 87     }
 88     if targData.includeWinners && overallWinner {
 89         keys[string(key)] = value
 90     }
 91 }
exchange/targeting.go
```
Finally the `BidderKey` function:
```
119 func (key TargetingKey) BidderKey(bidder BidderName, maxLength int) string {
120     s := string(key) + "_" + string(bidder)
121     if maxLength != 0 {
122         return s[:min(len(s), maxLength)]
123     }
124     return s
125 }
openrtb_ext/bid.go
```

camrice replied:
For the end to end test, I have a version of what you posted working now. The only issue I had is that I think I either need to add an unmarshal call in the test `HoldAuction` function to check if the flag is enabled (may not be a big deal performance wise since it is a unit test), or add a new `mockExchange` that adds the bidder name to the response which is how I have it implemented now.

For the second point, I see that this case is happening you mentioned where the key and value both have the bidder name added. Let me double check the impact of this today. The end result is as expected, but I think that I will be making `IncludeBidderKeys` and `AppendBidderNames` mutually exclusive. Doing this, I will have to add `IncludeWinners` back into the targeting values I believe.

guscarreon replied:
Awesome Cameron. Thank you for looking into this.

camrice replied:
No problem! I will add the end to end tests today. I just wanted to follow up on the second point after some discussion from today:

It seems that the functionality between the key and value for this feature is as expected for the Video endpoint. In the Video endpoint, the key name is fixed as `hb_pb_cat_dur` and we only want to update the value to have the bidder name. So in this case, this seems fine as is. The other question is whether this should be added to the Auction endpoint as well.

I will have to ask about that, but I was thinking that could be a separate PR if needed if that is okay with the team?

camrice replied:
Just another quick update: I will be scoping adding this to the Auction endpoint to see if it should be a separate PR or added to this one. After another discussion, it should be added to both endpoints.

guscarreon replied:
I'm fine if it's done on a separate PR as long as it is the approved functionality. I'll be on the look for any updates.

guscarreon reviewed (commented):
I believe it'd be very interesting, if possible, to write a test that asserts the targets in its final Json format. That is, the final jsonified `openrtb_ext.BidResponseVideo`. Unless it is too mmuch of a hassle, something like:
```
func TestAppendBidderNameToTargetEndToEnd(t *testing.T) {
    testCases := []struct {
        description             string
        bidRequestVideoJsonFile string
        expectedBidResponseExt  json.RawMessage
    }{
        {
            description:             "Valid video request with includeBidderKeys set to false",
            bidRequestVideoJsonFile: "sample-requests/video/video_valid_sample.json",
            expectedVideoBidRespExt: json.RawMessage(`{..."targeting": [{"hb_pb": "15.00","hb_pb_cat_dur": "15.00_394_30s", ... ]}`),
            // if this is the expected format -------------------------------------------------^
        },
        {
            description:             "Valid video request with includeBidderKeys set to true",
            bidRequestVideoJsonFile: "sample-requests/video/video_valid_sample_with_includeBidderKeys_true.json",
            expectedVideoBidRespExt: json.RawMessage(`{..."targeting": [{"hb_pb": "15.00","hb_pb_cat_dur": "15.00_394_30s_appnexus", ... ]}`),
            // if this is the expected format -------------------------------------------------^
        },
    }

    for _, test := range testCases {
        // Unmarshall test bidRequest from json file
        reqData, err := ioutil.ReadFile(test.bidRequestVideoJsonFile)
        if err != nil {
            t.Fatalf("Failed to fetch a valid request: %v", err)
        }
        reqBody := string(getRequestPayload(t, reqData))
        req := httptest.NewRequest("POST", "/openrtb2/video", strings.NewReader(reqBody))
        recorder := httptest.NewRecorder()

        // mock video endpoint with mock objects
        videoAuctionEndpoint, _ := NewVideoEndpoint(
            exchange.NewMockExchange(),
            newParamsValidator(t),
            &mockVideoStoredReqFetcher{},
            &mockVideoStoredReqFetcher{},
            empty_fetcher.EmptyFetcher{},
            empty_fetcher.EmptyFetcher{},
            &config.Configuration{MaxRequestSize: maxSize},
            pbsmetrics.NewMetrics(metrics.NewRegistry(), openrtb_ext.BidderList(), config.DisabledMetrics{}),
            &mockAnalyticsModule{},
            map[string]string{},
            []byte(`{"ext":{"prebid":{"aliases": {"test1": "appnexus", "test2": "rubicon", "test3": "openx"}}}}`),
            openrtb_ext.BidderMap,
            mockCacheClient{},
        )

        // Run test from end to end
        videoAuctionEndpoint(recorder, req, nil)

        // Assert actual bidVideoResponse vs expected
        assert.Equalf(t, test.expectedBidResponseExt, recorder.Body.Bytes() "Test failed")
    }
}
```

SyntaxNode commented:
The build failure is due to an unexpected change to our validation process. For now, please merge from the master branch to fix the issue.

camrice commented:
Sorry for the late update. I added an Video endpoint test and Exchange test for this change. I also fixed the unit test failure. If I should move my commits over to the Upstream repo and open a new PR let me know.

guscarreon reviewed (approved):
LGTM
